cpp/ql/src/semmle/code/cpp/valuenumbering/HashCons.qll

```diff
@@ -0,0 +1,519 @@
+/**
+ * Provides an implementation of Hash consing.
+ * See https://en.wikipedia.org/wiki/Hash_consing
+ *
+ * The predicate `hashCons` converts an expression into a `HC`, which is an
+ * abstract type presenting the hash-cons of the expression. If two
+ * expressions have the same `HC` then they are structurally equal.
+ *
+ * Important note: this library ignores the possibility that the value of
+ * an expression might change between one occurrence and the next. For
+ * example:
+ *
+ * ```
+ * x = a+b;
+ * a++;
+ * y = a+b;
+ * ```
+ *
+ * In this example, both copies of the expression `a+b` will hash-cons to
+ * the same value, even though the value of `a` has changed. This is the
+ * intended behavior of this library. If you care about the value of the
+ * expression being the same, then you should use the GlobalValueNumbering
+ * library instead.
+ *
+ * To determine if the expression `x` is structurally equal to the
+ * expression `y`, use the library like this:
+ *
+ * ```
+ * hashCons(x) = hashCons(y)
+ * ```
+ */
+
+/*
+ * Note to developers: the correctness of this module depends on the
+ * definitions of HC, hashCons, and analyzableExpr being kept in sync with
+ * each other. If you change this module then make sure that the change is
+ * symmetric across all three.
+ */
+
+import cpp
+
+/** Used to represent the hash-cons of an expression. */
+private cached newtype HCBase =
+  HC_IntLiteral(int val, Type t) { mk_IntLiteral(val,t,_) }
+  or
+  HC_FloatLiteral(float val, Type t) { mk_FloatLiteral(val,t,_) }
+  or
+  HC_StringLiteral(string val, Type t) {mk_StringLiteral(val,t,_)}
+  or
+  HC_Variable(Variable x) {
+     mk_Variable(x, _)
+  }
+  or
+  HC_FieldAccess(HC s, Field f) {
+    mk_DotFieldAccess(s,f,_) or
+    mk_PointerFieldAccess_with_deref(s,f,_) or
+    mk_ImplicitThisFieldAccess_with_deref(s,f,_)
+  }
+  or
+  HC_Deref(HC p) {
+    mk_Deref(p,_) or
+    mk_PointerFieldAccess(p,_,_) or
+    mk_ImplicitThisFieldAccess_with_qualifier(p,_,_)
+  }
+  or
+  HC_ThisExpr(Function fcn) {
+    mk_ThisExpr(fcn,_) or
+    mk_ImplicitThisFieldAccess(fcn,_,_)
+  }
+  or
+  HC_Conversion(Type t, HC child) { mk_Conversion(t, child, _) }
+  or
+  HC_BinaryOp(HC lhs, HC rhs, string opname) {
+    mk_BinaryOp(lhs, rhs, opname, _)
+  }
+  or
+  HC_UnaryOp(HC child, string opname) { mk_UnaryOp(child, opname, _) }
+  or
+  HC_ArrayAccess(HC x, HC i) {
+    mk_ArrayAccess(x,i,_)
+  }
+  or
+  HC_NonmemberFunctionCall(Function fcn, HC_Args args) {
+    mk_NonmemberFunctionCall(fcn, args, _)
+  }
+  or
+  HC_MemberFunctionCall(Function trg, HC qual, HC_Args args) {
+    mk_MemberFunctionCall(trg, qual, args, _)
+  }
+  or
+  // Any expression that is not handled by the cases above is
+  // given a unique number based on the expression itself.
+  HC_Unanalyzable(Expr e) { not analyzableExpr(e,_) }
+
+/** Used to implement hash-consing of argument lists  */
+private cached newtype HC_Args =
```

Comment:
Drive-by comment: When I looked into implementing a similar library for JavaScript, I got a considerable speed-up from adding special-case constructors for argument lists of length one and two. Have you investigated whether this makes a difference for C++?

Reply:
I haven't. I'll take a look at that, though.